packets2/packets2.go

```diff
@@ -0,0 +1,195 @@
+// Package packets2 implements MQTT-SN version 2.0 packets structs.
+package packets2
+
+import (
+	"fmt"
+	"io"
+
+	pkts "github.com/energomonitor/bisquitt/packets"
+)
+
+// MQTT-SN specification defines maximal packet size
+// to be 65535B but pion/udp and pion/dtls use maximal packet size of 8192B.
+// See e.g.:
+// - https://github.com/pion/udp/blob/b66c29020370bbb21647c27cf0b5ac50a18677f7/conn.go#L17
+// - https://github.com/pion/dtls/blob/3dc563b9aede91561ece5ae14b6ec6edf6fc5eb9/conn.go#L30
+// An effective MQTT-SN maximal packet size is even a few bytes smaller:
+// for UDP transport: PUBLISH with 8183B-long payload = 8192B total packet length
+// for DTLS transport: PUBLISH with 8146B-long payload = 8155B total packet length
+// (I'm not sure if the DTLS maximal length is affected by the cipher used or not)
+const MaxPacketLen = 8192
+
+type PacketV2 struct{}
+
+func (p PacketV2) MqttSnVersion() uint8 {
+	return 2
+}
+
+// Reason code constants.
+type ReasonCode uint8
+
+const (
+	RC_SUCCESS                      ReasonCode = 0x00
+	RC_CONGESTION                   ReasonCode = 0x01
+	RC_INVALID_TOPIC_ALIAS          ReasonCode = 0x02
+	RC_NOT_SUPPORTED                ReasonCode = 0x03
+	RC_NO_SESSION                   ReasonCode = 0x05
+	RC_UNSUPPORTED_PROTOCOL_VERSION ReasonCode = 0x84
+	RC_NOT_AUTHORIZED               ReasonCode = 0x87
+	RC_BAD_AUTHENTICATION_METHOD    ReasonCode = 0x8C
+	RC_PACKET_TOO_LARGE             ReasonCode = 0x95
+	RC_PAYLOAD_FORMAT_INVALID       ReasonCode = 0x99
+)
+
+func (rc ReasonCode) String() string {
+	switch rc {
+	case RC_SUCCESS:
+		return "success"
+	case RC_CONGESTION:
+		return "congestion"
+	case RC_INVALID_TOPIC_ALIAS:
+		return "invalid topic alias"
+	case RC_NOT_SUPPORTED:
+		return "not supported"
+	case RC_NO_SESSION:
+		return "no session"
+	case RC_UNSUPPORTED_PROTOCOL_VERSION:
+		return "unsupported protocol version"
+	case RC_NOT_AUTHORIZED:
+		return "not authorized"
+	case RC_BAD_AUTHENTICATION_METHOD:
+		return "bad authentication error"
+	case RC_PACKET_TOO_LARGE:
+		return "packet too large"
+	case RC_PAYLOAD_FORMAT_INVALID:
+		return "payload format invalid"
+	default:
+		return fmt.Sprintf("invalid (%d)", rc)
+	}
+}
+
+// Topic Alias Type constants.
+type TopicAliasType uint8
+
+const (
+	TAT_NORMAL TopicAliasType = iota
+	TAT_PREDEFINED
+	TAT_SHORT
+	TAT_LONG
+)
+
+func (tat TopicAliasType) String() string {
+	switch tat {
+	case TAT_NORMAL:
+		return "normal"
+	case TAT_PREDEFINED:
+		return "predefined"
+	case TAT_SHORT:
+		return "short"
+	case TAT_LONG:
+		return "long"
+	default:
+		return fmt.Sprintf("illegal(%d)", tat)
+	}
+}
+
+// Flags bits masks.
+const flagsNoLocalBit = 0b10000000
+const flagsDUPBit = 0b10000000
+const flagsQOSBits = 0b1100000
+const flagsRetainAsPublishedBit = 0b10000
+const flagsRetainBit = 0b10000
+const flagsRetainHandlingBits = 0b1100
+const flagsTopicAliasTypeBits = 0b11
```

Comment:
> I’d pad the consts with zeros

Makes sense, done.

> Also, use just one const statement with parens?

I wrote it this way to make it more explicit that it is not an "enum" (because Go doesn't have real enums, just consts...).

Reply:
> I wrote it this way to make it more explicit that it is not an "enum" (because Go doesn't have real enums, just consts...).

I see. I think it would be sufficiently clear even with one `const` statement but I don’t know enough about Go idioms to really tell. Leaving up to you.

Reply:
I think there's no idiom here. Varies with the author...